teamscale_precommit_client/precommit_client.py

```diff
@@ -8,94 +8,183 @@
 import argparse
 
 from teamscale_precommit_client.git_utils import get_current_branch, get_current_timestamp
-from teamscale_precommit_client.git_utils import get_changed_files_and_content, get_deleted_files
+from teamscale_precommit_client.git_utils import get_changed_files, get_changed_files_and_content, get_deleted_files
 from teamscale_precommit_client.data import PreCommitUploadData
 from teamscale_client import TeamscaleClient
 from teamscale_precommit_client.client_configuration_utils import get_teamscale_client_configuration
 from teamscale_precommit_client.git_utils import get_repo_root_from_file_in_repo
 
+# Filename of the precommit configuration. The client expects this config file at the root of the repository.
 PRECOMMIT_CONFIG_FILENAME = '.teamscale-precommit.config'
 
 
 class PrecommitClient:
     """Client for precommit analysis"""
+    # Number of seconds the client waits until fetching precommit results from the server.
+    PRECOMMIT_WAITING_TIME_IN_SECONDS = 2
 
     def __init__(self, teamscale_config, repository_path, analyzed_file=None, verify=True,
-                 omit_links_to_findings=False):
+                 omit_links_to_findings=False, exclude_findings_in_changed_code=False, fetch_existing_findings=False,
+                 fetch_all_findings=False, fetch_existing_findings_in_changes=False, fail_on_red_findings=False):
         """Constructor"""
         self.teamscale_client = TeamscaleClient(teamscale_config.url, teamscale_config.username,
                                                 teamscale_config.access_token, teamscale_config.project_id, verify)
         self.repository_path = repository_path
         self.analyzed_file = analyzed_file
-        self.omit_links_to_findings=omit_links_to_findings
+        self.omit_links_to_findings = omit_links_to_findings
+        self.exclude_findings_in_changed_code = exclude_findings_in_changed_code
+        self.fetch_existing_findings = fetch_existing_findings
+        self.fetch_all_findings = fetch_all_findings
+        self.fetch_existing_findings_in_changes = fetch_existing_findings_in_changes
+        self.fail_on_red_findings = fail_on_red_findings
+        self.changed_files = {}
```

Comment:
We have many fields hier and all from changed_files onwards is mutable state of this object. I propose having data classes for the `PreCommitUpload` (changed_files, deleted_files) and the resulting `FindingsChurn` (added/removed/existing findings) Which are returned (e.g. by `self._calculate_modifications()`) and passed through the functions as parameters. It'd be great to also do this for `findings_in_changed_code` as well as `current_branch` and `parent_commit_timestamp` if possible to reduce the mutable object state as well. Not sure if this leads to too many parameters passed through though. Maybe these fields could also be combined to some kind of PreCommitRequest (`findings_churn`, `current_branch`, `parent_commit_timestamp`) that can be executed to avoid passing many parameters.

Reply:
I like the idea of the data classes. However, I actually refactored the precommit client class to hold all that state, as before there were dozens of parameters (many of them representing the arguments) floating around through all methods. This was not readable and hard to test. Let me see if I can come up with something in the middle. I'm thinking of a configuration or session class or similar.

Reply:
As discussed, let's spawn a maintenance ticket to improve that.